gpjax/natural_gradients.py

```diff
@@ -0,0 +1,288 @@
+            params = transform(params, transformations)
+
+            # Get natural moments θ.
+            natural_moments = params["variational_family"]["moments"]
+
+            # Get expectation moments η.
+            expectation_moments = natural_to_expectation(natural_moments)
+
+            # Full params with expectation moments.
+            expectation_params = deepcopy(params)
+            expectation_params["variational_family"]["moments"] = expectation_moments
+
+            # Compute gradient ∂L/∂η:
+            def loss_fn(params: dict, batch: Dataset) -> f64["1"]:
+                # Determine hyperparameters that should be trained.
+                trains = deepcopy(trainables)
+                trains["variational_family"]["moments"] = build_trainables_true(
+                    params["variational_family"]["moments"]
+                )
+                params = trainable_params(params, trains)
+
+                # Stop gradients for non-moment parameters.
+                params = _stop_gradients_nonmoments(params)
+
+                return expectation_elbo(params, batch)
+
+            value, dL_dexp = value_and_grad(loss_fn)(expectation_params, batch)
+
+            # This is a renaming of the gradient components to match the natural parameterisation pytree.
+            nat_grad = dL_dexp
+            nat_grad["variational_family"]["moments"] = {
+                "natural_vector": dL_dexp["variational_family"]["moments"][
+                    "expectation_vector"
+                ],
+                "natural_matrix": dL_dexp["variational_family"]["moments"][
+                    "expectation_matrix"
+                ],
+            }
+
+            return value, nat_grad
+
+    else:
+        raise NotImplementedError
+
+    def hyper_grads_fn(params: dict, trainables: dict, batch: Dataset) -> dict:
+        """
+        Computes the hyperparameter gradients of the ELBO.
+        Args:
+            params: A dictionary of parameters.
+            trainables: A dictionary of trainables.
+            batch: A Dataset.
+        Returns:
+            dict: A dictionary of hyperparameter gradients.
+        """
+
+        def loss_fn(params: dict, batch: Dataset) -> f64["1"]:
+            # Determine hyperparameters that should be trained.
+            params = trainable_params(params, trainables)
+
+            # Stop gradients for the moment parameters.
+            params = _stop_gradients_moments(params)
+
+            return xi_elbo(params, batch)
+
+        value, dL_dhyper = value_and_grad(loss_fn)(params, batch)
+
+        return value, dL_dhyper
+
+    return nat_grads_fn, hyper_grads_fn
+
+
+def fit_natgrads(
+    stochastic_vi: StochasticVI,
+    params: tp.Dict,
+    trainables: tp.Dict,
+    transformations: tp.Dict,
+    train_data: Dataset,
+    batch_size: int,
+    moment_optim,
+    hyper_optim,
+    key: PRNGKeyType,
+    n_iters: tp.Optional[int] = 100,
+    log_rate: tp.Optional[int] = 10,
+) -> tp.Dict:
+
```

Comment:
We should have a docstring here explaining why we optimise the variational distribution then the hyperparameters i.e., stale posteriors. Also happy for this information to be duplicated in the notebook.